### PR TITLE
[FLINK-15632][kubernetes] Make zookeeper HA service could work for active kubernetes integration

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

Currently, when we deploy Flink cluster natively on Kubernetes and HA enabled, it could not work because of following two limitations
1. Flink client could not connect to the jobmanager when HA enabled. It will get a internal address that could not be accessed outside of Kubernetes cluster.
2. `JobManagerOptions#ADDRESS` has been set to service address when HA enabled. It has a risk  that the TaskManager will register to a wrong jobmanager when two jobmanagers are running. We need to use the pod ip address instead.

## Brief change log

* `hotfix` to make test could work for multiple pod environments.
* Always use Kubernetes service for client to contact with jobmanager in `KubernetesClusterDescriptor`
* Set ip address to `jobmanager.rpc.address` in `ClusterEntrypoint`
* Update `HighAvailabilityOptions.HA_CLUSTER_ID` and `HighAvailabilityOptions.HA_JOB_MANAGER_PORT_RANGE` when HA enabled

## Verifying this change

* Add a new unit test `testDeployHighAvailabilitySessionCluster`
* Test on a real Kubernetes cluster with zookeeper HA

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by wangyang0918** [comment]
@zhuzhurk Could you please take a look?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.flink.configuration.JobManagerOptions;
+import org.apache.flink.configuration.RestOptions;
+import org.apache.flink.kubernetes.utils.Constants;
+import org.apache.flink.runtime.jobmanager.HighAvailabilityMode;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * This class contains utility methods for the {@link KubernetesSessionClusterEntrypoint}.
+ */
+class KubernetesEntrypointUtils {
```
Comment (zhuzhurk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java]: It's better to add a private constructor for this util class

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		final String configDir = System.getenv(ConfigConstants.ENV_FLINK_CONF_DIR);
+		Preconditions.checkNotNull(
+			configDir,
+			"Flink configuration directory (%s) in environment should not be null!",
+			ConfigConstants.ENV_FLINK_CONF_DIR);
+
+		final Configuration configuration = GlobalConfiguration.loadConfiguration(configDir);
+
+		if (HighAvailabilityMode.isHighAvailabilityModeActivated(configuration)) {
+			final String ipAddress =  System.getenv().get(Constants.ENV_FLINK_POD_IP_ADDRESS);
```
Comment (zhuzhurk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java]:
```suggestion
			final String ipAddress = System.getenv().get(Constants.ENV_FLINK_POD_IP_ADDRESS);
```

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/FlinkMasterDeploymentDecorator.java [hunk 3]
```diff
@@ -138,13 +144,21 @@ private Container createJobManagerContainer(
 				new ContainerPortBuilder().withContainerPort(flinkConfig.getInteger(RestOptions.PORT)).build(),
 				new ContainerPortBuilder().withContainerPort(flinkConfig.getInteger(JobManagerOptions.PORT)).build(),
 				new ContainerPortBuilder().withContainerPort(blobServerPort).build()))
-			.withEnv(
-				BootstrapTools.getEnvironmentVariables(ResourceManagerOptions.CONTAINERIZED_MASTER_ENV_PREFIX, flinkConfig)
-					.entrySet()
-					.stream()
-					.map(kv -> new EnvVar(kv.getKey(), kv.getValue(), null))
-					.collect(Collectors.toList()))
+			.withEnv(buildEnvForContainer(flinkConfig)).withEnvFrom()
```
Comment (zhuzhurk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/FlinkMasterDeploymentDecorator.java]: useless `withEnvFrom`?

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8ClientTest.java [hunk 4]
```diff
@@ -189,7 +190,10 @@ public void testCreateFlinkMasterDeployment() {
 			jmContainer.getVolumeMounts().get(0).getMountPath());
 		assertEquals(FLINK_CONF_FILENAME, jmContainer.getVolumeMounts().get(0).getSubPath());
 
-		assertThat(jmContainer.getEnv(), Matchers.contains(new EnvVar(FLINK_MASTER_ENV_KEY, FLINK_MASTER_ENV_VALUE, null)));
+		EnvVar masterEnv = new EnvVar(FLINK_MASTER_ENV_KEY, FLINK_MASTER_ENV_VALUE, null);
```
Comment (zhuzhurk) [inline flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8ClientTest.java]: This change seems to be unrelated? How about to do it in a separate commit?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+			Preconditions.checkState(
+				ipAddress != null,
+				"JobManager ip address environment variable %s not set",
+				Constants.ENV_FLINK_POD_IP_ADDRESS);
+			configuration.setString(JobManagerOptions.ADDRESS, ipAddress);
```
Comment (zhuzhurk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java]: How about to set it to IP in all cases to avoid confusion, namely, change how we set `JobManagerOptions.ADDRESS` in `KubernetesClusterDescriptor#deployClusterInternal()`?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java [hunk 6]
```diff
@@ -84,6 +86,11 @@ public String getClusterDescription() {
 			if (restEndpoint != null) {
 				configuration.setString(RestOptions.ADDRESS, restEndpoint.getAddress());
 				configuration.setInteger(RestOptions.PORT, restEndpoint.getPort());
+				// Flink client will not use HA service to retrieve address of jobmanager. Instead, it always use Kubernetes
+				// service to contact with jobmanager via rest client.
+				if (HighAvailabilityMode.isHighAvailabilityModeActivated(flinkConfig)) {
```
Comment (zhuzhurk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: This might work. However, the way is a bit hack and add some dependency/assumption to the behavior of non-HA mode.
Is there a clean way we let the client retrieve jobmanager address from k8s service no matter what the HA mode is?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java (hunk 1 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java]: Nice suggestion. I will fix it.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/FlinkMasterDeploymentDecorator.java (hunk 3 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/FlinkMasterDeploymentDecorator.java]: Aha, i forget to remove it and will fix it.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8ClientTest.java (hunk 4 above)
Comment (wangyang0918) [reply flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8ClientTest.java]: Make sense. I will put it into a separate commit.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java (hunk 5 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesEntrypointUtils.java]: In non-HA mode, we need to set the jobmanager rpc address to service name. So when the jobmanager failover, the taskmanager could register again.
In HA mode, the taskmanager uses zookeeper to retrieve the jobmanager address. And there will be multiple jobmanagers in the future. So we set the ip address instead of jobmanager rpc address.

So i think we could not always set `JobManagerOptions.ADDRESS` to ip address.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java (hunk 6 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: I do not think it is a temporary hack. 
In Kubernetes deployment, when we want to get a ClusterClient, we should not retrieve it from the `ClientHAService`. Instead, we retrieve the rest endpoint from Kubernetes service. 

Maybe removing the condition `if (HighAvailabilityMode.isHighAvailabilityModeActivated(flinkConfig))` will look better. 

```
final Endpoint restEndpoint = client.getRestEndpoint(clusterId);
if (restEndpoint != null) {
  configuration.setString(RestOptions.ADDRESS, restEndpoint.getAddress());
  configuration.setInteger(RestOptions.PORT, restEndpoint.getPort());
}
```

Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: I have a serious consideration and will find a more clean way. Maybe adding a new construct method for `RestClusterClient` could work.

**Comment by wangyang0918** [comment]
@zhuzhurk I have addressed all your comments. Could you please take a look again?

File: flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java [hunk 7]
```diff
@@ -182,7 +206,7 @@ public RestClusterClient(Configuration config, T clusterId) throws Exception {
 		this.waitStrategy = checkNotNull(waitStrategy);
 		this.clusterId = checkNotNull(clusterId);
 
-		this.clientHAServices = HighAvailabilityServicesUtils.createClientHAService(configuration);
+		this.clientHAServices = clientHAServices;
```
Comment (zhuzhurk) [inline flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java]: better to `checkNotNull`

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java [hunk 8]
```diff
@@ -168,6 +185,13 @@ public String getClusterDescription() {
 		final String nameSpace = flinkConfig.getString(KubernetesConfigOptions.NAMESPACE);
 		flinkConfig.setString(JobManagerOptions.ADDRESS, clusterId + "." + nameSpace);
 
+		if (HighAvailabilityMode.isHighAvailabilityModeActivated(flinkConfig)) {
+			flinkConfig.setString(HighAvailabilityOptions.HA_CLUSTER_ID, clusterId);
```
Comment (zhuzhurk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: What's the config overriding for?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java [hunk 9]
```diff
@@ -91,7 +104,11 @@ public String getClusterDescription() {
 			}
 
 			try {
-				return new RestClusterClient<>(configuration, clusterId);
+				return new RestClusterClient<>(
+					configuration,
+					clusterId,
+					new StandaloneClientHAServices(HighAvailabilityServicesUtils.getWebMonitorAddress(
```
Comment (zhuzhurk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: It would be better to leave a comment here for why we use StandaloneClientHAServices for k8s client.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java (hunk 8 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: We should update the `HA_CLUSTER_ID` since it will be used to contruct the zk path and ha storage path for the Flink cluster.
Yarn has the similar behavior in `FlinkYarnSessionCli#applyCommandLineOptionsToConfiguration`.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java (hunk 9 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: I will add some comments here.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java (hunk 8 above)
Comment (zhuzhurk) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: So this change is not related to this commit?

Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: It is necessary when high-availability enabled. So do you mean to separate it to another commit?

Comment (zhuzhurk) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java]: Yes. If it fixes another issue, we should make it a different commit.

**Comment by wangyang0918** [comment]
@flinkbot run azure

**Comment by wangyang0918** [comment]
@zhuzhurk Thanks for the second round of review. I have addressed all your comments and please take a look at your convenience.

**Comment by zhuzhurk** [comment]
Thanks for addressing the comments @wangyang0918 .
The change looks good to me. Will merge it once travis gives green.

**Comment by wangyang0918** [comment]
@flinkbot run travis
